### ISSUE TITLE
Race condition in type inference when combined with record inheritance

### ISSUE THREAD

**Issue by jaredpar** [dotnet/roslyn#50040]
Consider the following code sample: 

```c#
#nullable enable
using System;
using System.Linq.Expressions;
using System.Threading.Tasks;

var collection = new Collection<Hamster>();
Hamster h = null!;
await collection.FooAsync(
    entity => entity.Name! == "bar",
    h);

public record Collection<T> 
    where T : Document
{
    public Task FooAsync<TDerived>(
        Expression<Func<TDerived, bool>> filter,
        TDerived td)
        where TDerived : T
    {
        throw new NotImplementedException();
    }

    public Task FooAsync<TDerived>(
        Task<TDerived> filterDefinition,
        TDerived td)
        where TDerived : T
    {
        throw new NotImplementedException();
    }
}

public sealed record HamsterCollection : Collection<Hamster>
{
}

public abstract class Document
{

}

public sealed class Hamster : Document
{
    public string? Name { get; private set; }
}
```

Compiling this code with .NET 5 SDK will succeed roughly 50% of the time. When the compilation fails the reported error is inability to infer the type of the following expression:

```c#
    entity => entity.Name! == "bar",
```

Now delete the `HamsterCollection` type in it's entirety and the compilation will succeed 100% of the time. It also succeeds if you change the declaration from `record` to `class`. There appears to be some race condition with records and inheritance here. 

This is a reduced sample from a bug report in [this tweet](https://twitter.com/i3arnon/status/1339631221052194820)

**Comment by jaredpar** [comment]
@jcouv FYI

**Comment by jcouv** [comment]
Summary of investigation: 
When multiple threads call `GetMembersByNameSlow` on a record type, we add synthesized members (`AddSynthesizedRecordMembersIfNecessary`) but this realizes the parameters of the methods (to compare signatures) which causes the binder cache to store methods which have not yet been reified as the official members of the type. 
If done concurrently, a second identical method is instantiated whose parameter references the type symbol of the first instance instead of its own.
This results in a broken object structure (duplicate method instance survives without being a member of its containing type, so doesn't get properly completed).
This causes assertions in debug mode and likely the above symptoms in release mode (I will confirm after fixing).


References:
- caching of method symbols in binder: [BinderFactory.VisitMethodDeclaration](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs#L131)
- the method symbol that is cached is `this` method when binding parameters: [SourceOrdinaryMethodSymbol.MakeParametersAndBindReturnType](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs#L123)
- the stacktrace below shows how the comparison of symbols while computing members of records realizes parameters:

```
   at Microsoft.CodeAnalysis.CSharp.BinderFactory.BinderFactoryVisitor.Initialize(Int32 position, CSharpSyntaxNode memberDeclarationOpt, Symbol memberOpt)
   at Microsoft.CodeAnalysis.CSharp.BinderFactory.GetBinder(SyntaxNode node, Int32 position, CSharpSyntaxNode memberDeclarationOpt, Symbol memberOpt)
   at Microsoft.CodeAnalysis.CSharp.BinderFactory.GetBinder(SyntaxNode node, CSharpSyntaxNode memberDeclarationOpt, Symbol memberOpt)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceOrdinaryMethodSymbol.MakeParametersAndBindReturnType(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceOrdinaryMethodSymbolBase.MethodChecks(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberMethodSymbol.LazyMethodChecks()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceOrdinaryMethodSymbolBase.get_Parameters()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SymbolExtensions.GetParameters(Symbol member)
   at Microsoft.CodeAnalysis.CSharp.Symbols.MemberSignatureComparer.Equals(Symbol member1, Symbol member2)
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at System.Collections.Generic.Dictionary`2.ContainsKey(TKey key)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.AddSynthesizedRecordMembersIfNecessary(MembersAndInitializersBuilder builder, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.BuildMembersAndInitializers(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.GetMembersAndInitializers()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.MakeAllMembers(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.GetMembersByNameSlow()
```

**Comment by jcouv** [comment]
Fixed in 16.9p4
